pkg/controllers/machinesetsync/machineset_sync_controller.go

```diff
@@ -791,53 +772,299 @@ func (r *MachineSetSyncReconciler) createOrUpdateCAPIInfraMachineTemplate(ctx co
 
-	if capiMachineSet == nil {
-		if err := r.Create(ctx, newCAPIMachineSet); err != nil {
-			logger.Error(err, "Failed to create CAPI machine set")
+	// If there is no existing CAPI machine set, create a new one.
+	if created, err := r.ensureCAPIMachineSet(ctx, mapiMachineSet, existingCAPIMachineSet, convertedCAPIMachineSet); err != nil {
+		return fmt.Errorf("failed to ensure CAPI machine set: %w", err)
+	} else if created {
+		// If the CAPI machine set was created,
+		// return early to allow for the change to be propagated.
+		return nil
+	}
 
-			createErr := fmt.Errorf("failed to create CAPI machine set: %w", err)
-			if condErr := r.applySynchronizedConditionWithPatch(
-				ctx, mapiMachineSet, corev1.ConditionFalse, reasonFailedToCreateCAPIMachineSet, createErr.Error(), nil); condErr != nil {
-				return utilerrors.NewAggregate([]error{createErr, condErr})
-			}
+	// If there is an existing CAPI machine set, work out if it needs updating.
 
-			return createErr
-		}
+	// Compare the existing CAPI machine set with the converted CAPI machine set to check for changes.
+	capiMachineSetsDiff := compareCAPIMachineSets(existingCAPIMachineSet, convertedCAPIMachineSet)
 
-		logger.Info("Successfully created CAPI machine set", "name", newCAPIMachineSet.Name, "infraMachineTemplate", newCAPIMachineSet.Spec.Template.Spec.InfrastructureRef.Name)
+	// Make a deep copy of the converted CAPI machine set to avoid modifying the original.
+	updatedOrCreatedCAPIMachineSet := convertedCAPIMachineSet.DeepCopy()
 
-		return nil
+	// If there are spec changes, update the CAPI machine set.
+	specUpdated, err := r.ensureCAPIMachineSetSpecUpdated(ctx, mapiMachineSet, capiMachineSetsDiff, updatedOrCreatedCAPIMachineSet)
+	if err != nil {
+		return fmt.Errorf("failed to ensure CAPI machine set spec updated: %w", err)
 	}
 
-	capiMachineSetsDiff := compareCAPIMachineSets(capiMachineSet, newCAPIMachineSet)
+	// If there are status changes, update the CAPI machine set status.
+	statusUpdated, err := r.ensureCAPIMachineSetStatusUpdated(ctx, mapiMachineSet, existingCAPIMachineSet, convertedCAPIMachineSet, updatedOrCreatedCAPIMachineSet, capiMachineSetsDiff, specUpdated)
+	if err != nil {
+		return fmt.Errorf("failed to ensure CAPI machine set status updated: %w", err)
+	}
 
-	if len(capiMachineSetsDiff) == 0 {
+	if specUpdated || statusUpdated {
+		logger.Info("Successfully updated CAPI machine set")
+	} else {
 		logger.Info("No changes detected for CAPI machine set")
-		return nil
+	}
+
+	return nil
+}
+
+// ensureCAPIMachineSet creates a new CAPI machine set if one doesn't exist.
+func (r *MachineSetSyncReconciler) ensureCAPIMachineSet(ctx context.Context, sourceMAPIMachineSet *machinev1beta1.MachineSet, existingCAPIMachineSet *clusterv1.MachineSet, convertedCAPIMachineSet *clusterv1.MachineSet) (bool, error) {
+	// If there is an existing CAPI machine set, no need to create one.
+	if existingCAPIMachineSet != nil {
+		return false, nil
+	}
```

Comment:
This is the only `return false, nil` case. Should we call this func `createCAPIMachineSet` and move this comparison to the above function?

This makes it more readable.

Reply:
It used to be like that, but I switched to this approach. I like the "ensure" style naming as to me it hides all the checking logic, whatever it is, and ensures after that I have what I expect.

Also it lowers complexity of the outer function, line count and gocognit

Reply:
In other codebases I've seen, ensure generally means check this thing exists, and that it looks like a certain shape. So it would make sense in the context of createOrUpdate, but maybe is less aligned with this usage where it is just creating, with updates handled separately?